### DISCUSSION TITLE
UIUC CS513: Theory and Practice of Data Cleaning follow up discussion

### DISCUSSION THREAD

**Discussion by GhosanandaW** [EvgSkv/logica#351] [General]
Hi Dr. Skvortsov and Logica developer team, 

Thank you very much for the very insightful sharing given this week! We discussed Logica's capability on "data type checking" during the Q&A question and I would like to further deep dive on the matter here if that's all right.

Since we are trying to use Logica as denial constraint and inclusion constraint implementation in the data cleaning and database, I believe a data type checking would be one of the great way to achieve this use case. 

For example, if we are ensuring a Foreign Key from attribute "cities" in table A that has a datatype of list of string of multiple cities (e.g., ["Singapore", "Tokyo", "San Diego"]), compared with attribute "cities" in table B (that supposedly also a list of string too), but in one of the tuples it is inserted as string of multiple cities (e.g., "Singapore", "Tokyo", "San Diego"). The mismatch of datatype detection will help as a second layer after the schema check/trigger constraint in the DBMS. What would be the best way to implement a detection/warning of the type mismatch in Logica to ensure the constraints?

Thank you very much for the time and help in advance!

**Comment by EvgSkv** [comment]
Hello @GhosanandaW 
Logica's type checking is similar to type checking in Python, C++, etc: it checks whether user had incorrect types in the program and produes an immediate error.

Usually you assume that you know types in the tables and you need to check that content doesn't violate constaraints. So I am not sure how could typechecking help here...

But maybe if you have types described as data itself, then we can use logica to describe constraints more logically. But it won't be using logica type checking feature.

Hope it helps, let me know if you have further questions!

**Reply by GhosanandaW** [reply]
Thank you very much for the information given Dr. Skvortsov.

I also noticed that DuckDB support has been added to Logica. Would there be any reference documentation we can refer to for the setup?

**Reply by EvgSkv** [reply]
Yes, that's right!

Here is colab for reference on how to use DuckDB:
https://colab.research.google.com/drive/1LCqmUocOnJ9gjFIo3GYWrRheiRIF9ilt?usp=sharing

Please let me know if you have further questions!
